File: frontend/lib/avatarUtils.ts
/**
 * Avatar utility functions for deterministic avatar generation.
 *
 * Features:
 * - Extract initials from first and last name (JD for "John Doe")
 * - Generate deterministic colors from user ID or name
 * - WCAG AA compliant color contrast
 * - Handles edge cases (single name, non-Latin characters, empty names)
 */

/**
 * WCAG AA compliant color palette for avatar backgrounds.
 * Each color has a contrast ratio of at least 4.5:1 with white text.
 */
export const AVATAR_COLORS = [
  { bg: '#6366f1', text: '#ffffff' }, // Indigo
  { bg: '#8b5cf6', text: '#ffffff' }, // Violet
  { bg: '#a855f7', text: '#ffffff' }, // Purple
  { bg: '#d946ef', text: '#ffffff' }, // Fuchsia
  { bg: '#ec4899', text: '#ffffff' }, // Pink
  { bg: '#f43f5e', text: '#ffffff' }, // Rose
  { bg: '#ef4444', text: '#ffffff' }, // Red
  { bg: '#f97316', text: '#ffffff' }, // Orange
  { bg: '#eab308', text: '#1f2937' }, // Yellow (dark text)
  { bg: '#84cc16', text: '#1f2937' }, // Lime (dark text)
  { bg: '#22c55e', text: '#ffffff' }, // Green
  { bg: '#10b981', text: '#ffffff' }, // Emerald
  { bg: '#14b8a6', text: '#ffffff' }, // Teal
  { bg: '#06b6d4', text: '#ffffff' }, // Cyan
  { bg: '#0ea5e9', text: '#ffffff' }, // Sky
  { bg: '#3b82f6', text: '#ffffff' }, // Blue
] as const;

/**
 * Tailwind-compatible color classes for avatar backgrounds.
 * Maps to AVATAR_COLORS indices for consistency.
 */
export const AVATAR_COLOR_CLASSES = [
  { bg: 'bg-indigo-500', text: 'text-white' },
  { bg: 'bg-violet-500', text: 'text-white' },
  { bg: 'bg-purple-500', text: 'text-white' },
  { bg: 'bg-fuchsia-500', text: 'text-white' },
  { bg: 'bg-pink-500', text: 'text-white' },
  { bg: 'bg-rose-500', text: 'text-white' },
  { bg: 'bg-red-500', text: 'text-white' },
  { bg: 'bg-orange-500', text: 'text-white' },
  { bg: 'bg-yellow-500', text: 'text-gray-800' },
  { bg: 'bg-lime-500', text: 'text-gray-800' },
  { bg: 'bg-green-500', text: 'text-white' },
  { bg: 'bg-emerald-500', text: 'text-white' },
  { bg: 'bg-teal-500', text: 'text-white' },
  { bg: 'bg-cyan-500', text: 'text-white' },
  { bg: 'bg-sky-500', text: 'text-white' },
  { bg: 'bg-blue-500', text: 'text-white' },
] as const;

/**
 * Simple hash function for deterministic color selection.
 * Uses djb2 algorithm for good distribution.
 */
export function hashString(str: string): number {
  let hash = 5381;
  for (let i = 0; i < str.length; i++) {
    hash = ((hash << 5) + hash) ^ str.charCodeAt(i);
  }
  return Math.abs(hash);
}

/**
 * Get deterministic color index from user ID or name.
 * Same input always produces the same color.
 *
 * @param identifier - User ID (number) or name/email (string)
 * @returns Index into AVATAR_COLORS array
 */
export function getAvatarColorIndex(identifier: string | number): number {
  const hash = typeof identifier === 'number'
    ? identifier
    : hashString(String(identifier).toLowerCase());
  return hash % AVATAR_COLORS.length;
}

/**
 * Get avatar color by identifier (user ID or name).
 * Returns both hex values and Tailwind classes.
 *
 * @param identifier - User ID or name
 * @returns Object with hex colors and Tailwind classes
 */
export function getAvatarColor(identifier: string | number) {
  const index = getAvatarColorIndex(identifier);
  return {
    hex: AVATAR_COLORS[index],
    classes: AVATAR_COLOR_CLASSES[index],
    index,
  };
}

/**
 * Extract initials from a name.
 *
 * Rules:
 * 1. "John Doe" → "JD" (first letter of first and last name)
 * 2. "John" → "JO" (first two letters if single name)
 * 3. "J" → "J" (single character stays as-is)
 * 4. Non-Latin characters are supported as-is
 * 5. Empty/invalid → "??"
 *
 * @param name - Full name or single name
 * @returns 1-2 character initials, uppercase
 */
export function getInitials(name: string | null | undefined): string {
  if (!name || typeof name !== 'string') {
    return '??';
  }

  // Trim and normalize whitespace
  const trimmed = name.trim().replace(/\s+/g, ' ');

  if (trimmed.length === 0) {
    return '??';
  }

  // Split by spaces to get name parts
  const parts = trimmed.split(' ').filter(part => part.length > 0);

  if (parts.length === 0) {
    return '??';
  }

  if (parts.length === 1) {
    // Single name: use first two characters
    const singleName = parts[0];
    if (singleName.length === 1) {
      return singleName.toUpperCase();
    }
    return singleName.slice(0, 2).toUpperCase();
  }

  // Multiple names: first letter of first and last name
  const firstName = parts[0];
  const lastName = parts[parts.length - 1];

  const firstInitial = firstName.charAt(0);
  const lastInitial = lastName.charAt(0);

  return (firstInitial + lastInitial).toUpperCase();
}

/**
 * Check if a URL is a placeholder/default avatar that should be ignored.
 */
export function isPlaceholderUrl(url: string | null | undefined): boolean {
  if (!url) return true;

  const placeholderPatterns = [
    'ui-avatars.com',
    'placehold.co',
    'placeholder.com',
    'via.placeholder.com',
    'dummyimage.com',
    'placekitten.com',
    'picsum.photos',
  ];

  return placeholderPatterns.some(pattern => url.includes(pattern));
}

/**
 * Generate alt text for an avatar.
 *
 * @param name - User's name
 * @returns Accessible alt text
 */
export function getAvatarAltText(name: string | null | undefined): string {
  if (!name || typeof name !== 'string' || name.trim().length === 0) {
    return 'User avatar';
  }
  return `Avatar for ${name.trim()}`;
}

export interface AvatarGenerationResult {
  initials: string;
  backgroundColor: string;
  textColor: string;
  bgClass: string;
  textClass: string;
  altText: string;
  colorIndex: number;
}

/**
 * Generate all avatar properties from a name and optional identifier.
 * Use this for complete avatar generation in one call.
 *
 * @param name - User's display name
 * @param identifier - Optional user ID for more stable color assignment
 * @returns Complete avatar generation result
 */
export function generateAvatar(
  name: string | null | undefined,
  identifier?: string | number
): AvatarGenerationResult {
  const colorKey = identifier ?? name ?? 'unknown';
  const color = getAvatarColor(colorKey);

  return {
    initials: getInitials(name),
    backgroundColor: color.hex.bg,
    textColor: color.hex.text,
    bgClass: color.classes.bg,
    textClass: color.classes.text,
    altText: getAvatarAltText(name),
    colorIndex: color.index,
  };
}

/**
 * Get CSS style object for inline avatar styling.
 * Useful when Tailwind classes aren't available.
 *
 * @param name - User's display name
 * @param identifier - Optional user ID
 * @param size - Avatar size in pixels
 * @returns CSS properties object
 */
export function getAvatarStyle(
  name: string | null | undefined,
  identifier?: string | number,
  size: number = 48
): React.CSSProperties {
  const { backgroundColor, textColor, initials } = generateAvatar(name, identifier);

  return {
    width: size,
    height: size,
    borderRadius: '50%',
    backgroundColor,
    color: textColor,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    fontWeight: 600,
    fontSize: size * 0.4,
    textTransform: 'uppercase' as const,
  };
}

File: backend/core/avatar_utils.py
"""
Avatar utility functions for deterministic avatar generation.

This module provides backend utilities for avatar generation that match
the frontend implementation for consistency across the platform.

Features:
- Extract initials from first and last name (JD for "John Doe")
- Generate deterministic colors from user ID or name
- WCAG AA compliant color contrast
- Handles edge cases (single name, non-Latin characters, empty names)
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# WCAG AA compliant color palette for avatar backgrounds.
# Each color has a contrast ratio of at least 4.5:1 with its text color.
AVATAR_COLORS: list[tuple[str, str]] = [
    ("#6366f1", "#ffffff"),  # Indigo
    ("#8b5cf6", "#ffffff"),  # Violet
    ("#a855f7", "#ffffff"),  # Purple
    ("#d946ef", "#ffffff"),  # Fuchsia
    ("#ec4899", "#ffffff"),  # Pink
    ("#f43f5e", "#ffffff"),  # Rose
    ("#ef4444", "#ffffff"),  # Red
    ("#f97316", "#ffffff"),  # Orange
    ("#eab308", "#1f2937"),  # Yellow (dark text)
    ("#84cc16", "#1f2937"),  # Lime (dark text)
    ("#22c55e", "#ffffff"),  # Green
    ("#10b981", "#ffffff"),  # Emerald
    ("#14b8a6", "#ffffff"),  # Teal
    ("#06b6d4", "#ffffff"),  # Cyan
    ("#0ea5e9", "#ffffff"),  # Sky
    ("#3b82f6", "#ffffff"),  # Blue
]


def hash_string(s: str) -> int:
    """
    Simple hash function for deterministic color selection.
    Uses djb2 algorithm for good distribution.
    Matches the frontend implementation exactly.
    """
    hash_value = 5381
    for char in s:
        hash_value = ((hash_value << 5) + hash_value) ^ ord(char)
    return abs(hash_value)


def get_avatar_color_index(identifier: str | int) -> int:
    """
    Get deterministic color index from user ID or name.
    Same input always produces the same color.

    Args:
        identifier: User ID (int) or name/email (str)

    Returns:
        Index into AVATAR_COLORS array
    """
    hash_value = identifier if isinstance(identifier, int) else hash_string(str(identifier).lower())
    return hash_value % len(AVATAR_COLORS)


def get_avatar_color(identifier: str | int) -> tuple[str, str]:
    """
    Get avatar colors by identifier (user ID or name).

    Args:
        identifier: User ID or name

    Returns:
        Tuple of (background_color, text_color) as hex strings
    """
    index = get_avatar_color_index(identifier)
    return AVATAR_COLORS[index]


def get_initials(name: str | None) -> str:
    """
    Extract initials from a name.

    Rules:
    1. "John Doe" → "JD" (first letter of first and last name)
    2. "John" → "JO" (first two letters if single name)
    3. "J" → "J" (single character stays as-is)
    4. Non-Latin characters are supported as-is
    5. Empty/invalid → "??"

    Args:
        name: Full name or single name

    Returns:
        1-2 character initials, uppercase
    """
    if not name or not isinstance(name, str):
        return "??"

    # Trim and normalize whitespace
    trimmed = re.sub(r"\s+", " ", name.strip())

    if not trimmed:
        return "??"

    # Split by spaces to get name parts
    parts = [p for p in trimmed.split(" ") if p]

    if not parts:
        return "??"

    if len(parts) == 1:
        # Single name: use first two characters
        single_name = parts[0]
        if len(single_name) == 1:
            return single_name.upper()
        return single_name[:2].upper()

    # Multiple names: first letter of first and last name
    first_name = parts[0]
    last_name = parts[-1]

    first_initial = first_name[0] if first_name else ""
    last_initial = last_name[0] if last_name else ""

    return (first_initial + last_initial).upper()


def is_placeholder_url(url: str | None) -> bool:
    """
    Check if a URL is a placeholder/default avatar that should be ignored.

    Args:
        url: Avatar URL to check

    Returns:
        True if URL is a placeholder, False otherwise
    """
    if not url:
        return True

    placeholder_patterns = [
        "ui-avatars.com",
        "placehold.co",
        "placeholder.com",
        "via.placeholder.com",
        "dummyimage.com",
        "placekitten.com",
        "picsum.photos",
    ]

    return any(pattern in url for pattern in placeholder_patterns)


def get_avatar_alt_text(name: str | None) -> str:
    """
    Generate alt text for an avatar.

    Args:
        name: User's name

    Returns:
        Accessible alt text
    """
    if not name or not isinstance(name, str) or not name.strip():
        return "User avatar"
    return f"Avatar for {name.strip()}"


@dataclass
class AvatarData:
    """Complete avatar generation result."""

    initials: str
    background_color: str
    text_color: str
    alt_text: str
    color_index: int


def generate_avatar(
    name: str | None,
    identifier: str | int | None = None,
) -> AvatarData:
    """
    Generate all avatar properties from a name and optional identifier.
    Use this for complete avatar generation in one call.

    Args:
        name: User's display name
        identifier: Optional user ID for more stable color assignment

    Returns:
        Complete avatar generation result
    """
    color_key = identifier if identifier is not None else (name or "unknown")
    bg_color, text_color = get_avatar_color(color_key)
    color_index = get_avatar_color_index(color_key)

    return AvatarData(
        initials=get_initials(name),
        background_color=bg_color,
        text_color=text_color,
        alt_text=get_avatar_alt_text(name),
        color_index=color_index,
    )


def generate_avatar_svg(
    name: str | None,
    identifier: str | int | None = None,
    size: int = 100,
) -> str:
    """
    Generate an SVG avatar image.
    Useful for email templates, PDF generation, or server-side rendering.

    Args:
        name: User's display name
        identifier: Optional user ID for color stability
        size: SVG size in pixels

    Returns:
        SVG string
    """
    avatar = generate_avatar(name, identifier)
    font_size = int(size * 0.4)

    return f"""<svg xmlns="http://www.w3.org/2000/svg" width="{size}" height="{size}" viewBox="0 0 {size} {size}">
  <circle cx="{size // 2}" cy="{size // 2}" r="{size // 2}" fill="{avatar.background_color}"/>
  <text x="50%" y="50%" dominant-baseline="central" text-anchor="middle"
        fill="{avatar.text_color}" font-family="system-ui, -apple-system, sans-serif"
        font-size="{font_size}" font-weight="600">
    {avatar.initials}
  </text>
</svg>"""


def generate_avatar_data_uri(
    name: str | None,
    identifier: str | int | None = None,
    size: int = 100,
) -> str:
    """
    Generate a data URI for an avatar SVG.
    Can be used directly in img src attributes.

    Args:
        name: User's display name
        identifier: Optional user ID for color stability
        size: SVG size in pixels

    Returns:
        Data URI string
    """
    import urllib.parse

    svg = generate_avatar_svg(name, identifier, size)
    # Use URL encoding for better compatibility
    encoded = urllib.parse.quote(svg)
    return f"data:image/svg+xml,{encoded}"
